File: src/101. Symmetric Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @return {boolean}
 */

function isSymmetric(root) {
  if (root === null) return true
  return isSameNode(root.left, root.right)
}

function isSameNode(left, right) {
  if (left === null && right === null) return true
  if (left !== null && right === null) return false
  if (left === null && right !== null) return false
  if (left.val !== right.val) return false
  return (
    isSameNode(left.right, right.left) && isSameNode(left.left, right.right)
  )
}

File: src/101. Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool isSameNode(TreeNode* root1, TreeNode* root2) {
        if (!root1 && !root2) {
            return true;
        }
        if (!root1 || !root2 || root1->val != root2->val) {
            return false;
        }
        return isSameNode(root1->left, root2->right) && isSameNode(root1->right, root2->left);
    }
public:
    bool isSymmetric(TreeNode* root) {
        return isSameNode(root, root);
    }
};